snapcraft/models/project.py

```diff
@@ -436,13 +436,23 @@ def _validate_autostart_name(cls, name):
 
         return name
 
-    @pydantic.field_validator("bus_name")
+    @pydantic.field_validator(
+        "command", "stop_command", "post_stop_command", "reload_command", "bus_name"
+    )
     @classmethod
-    def _validate_bus_name(cls, name):
-        if not re.match(r"^[A-Za-z0-9/. _#:$-]*$", name):
-            raise ValueError(f"{name!r} is not a valid bus name")
-
-        return name
+    def _validate_apps_section_content(
+        cls, command: str, info: pydantic.ValidationInfo
+    ) -> str:
+        # Find any invalid characters in the field.
+        # The regex below is derived from snapd's validator code.
+        # https://github.com/canonical/snapd/blob/0706e2d0b20ae2bf030863f142b8491b66e80bcb/snap/validate.go#L756
+        if not re.match(r"^[A-Za-z0-9/. _#:$-]*$", command):
+            # Guaranteed not-none as the pydantic field_validator decorator always populates it.
+            deserialized = cast(str, info.field_name).replace("_", "-")
+            message = f"{deserialized}: App commands must consist of only alphanumeric characters, spaces, and the following characters: / . _ # : $ -"
```

Comment:
I think there is no need to parse a field name as it is already there as dotted path, e.g. in your example from different comment:
```
apps.myapp.command
  Value error, command: 
```
```suggestion
            message = f"app section element must consist of only alphanumeric characters, spaces, and the following characters: / . _ # : $ -"
```

Reply:
That's a good point, I agree. Thanks! I'll wait to see what @mr-cal says about the changelog - it'd be good for me to finally add to one.